beacon_node/rest_api/src/validator.rs

```diff
@@ -147,3 +130,147 @@ pub fn get_validator_duties<T: BeaconChainTypes + 'static>(req: Request<Body>) -
     );
     Ok(success_response(body))
 }
+
+/// HTTP Handler to produce a new BeaconBlock from the current state, ready to be signed by a validator.
+pub fn get_new_beacon_block<T: BeaconChainTypes + 'static>(req: Request<Body>) -> ApiResult {
+    let beacon_chain = get_beacon_chain_from_request::<T>(&req)?;
+
+    let query = UrlQuery::from_request(&req)?;
+    let slot = query
+        .first_of(&["slot"])
+        .map(|(_key, value)| value)?
+        .parse::<u64>()
+        .map(Slot::from)
+        .map_err(|e| {
+            ApiError::InvalidQueryParams(format!("Invalid slot parameter, must be a u64. {:?}", e))
+        })?;
+    let randao_bytes = query
+        .first_of(&["randao_reveal"])
+        .map(|(_key, value)| value)
+        .map(hex::decode)?
+        .map_err(|e| {
+            ApiError::InvalidQueryParams(format!("Invalid hex string for randao_reveal: {:?}", e))
+        })?;
+    let randao_reveal = Signature::from_bytes(randao_bytes.as_slice()).map_err(|e| {
+        ApiError::InvalidQueryParams(format!("randao_reveal is not a valid signature: {:?}", e))
+    })?;
+
+    let (new_block, _state) = beacon_chain
+        .produce_block(randao_reveal, slot)
+        .map_err(|e| {
+            ApiError::ServerError(format!(
+                "Beacon node is not able to produce a block: {:?}",
+                e
+            ))
+        })?;
+
+    let body = Body::from(
+        serde_json::to_string(&new_block)
+            .expect("We should always be able to serialize a new block that we produced."),
+    );
+    Ok(success_response(body))
+}
+
+/// HTTP Handler to produce a new Attestation from the current state, ready to be signed by a validator.
+pub fn get_new_attestation<T: BeaconChainTypes + 'static>(req: Request<Body>) -> ApiResult {
+    let beacon_chain = get_beacon_chain_from_request::<T>(&req)?;
+    let head_state = &beacon_chain.head().beacon_state;
+
+    let query = UrlQuery::from_request(&req)?;
+    let val_pk_str = query
+        .first_of(&["validator_pubkey"])
+        .map(|(_key, value)| value)?;
+    let val_pk = parse_pubkey(val_pk_str.as_str())?;
+
+    // Get the validator index from the supplied public key
+    // If it does not exist in the index, we cannot continue.
+    let val_index = head_state
+        .get_validator_index(&val_pk)
+        .map_err(|e| {
+            ApiError::ServerError(format!("Unable to read validator index cache. {:?}", e))
+        })?
+        .ok_or(ApiError::InvalidQueryParams(
+            "The provided validator public key does not correspond to a validator index.".into(),
+        ))?;
+
+    // Get the duties of the validator, to make sure they match up.
+    // If they don't have duties this epoch, then return an error
+    let val_duty = head_state
```

Comment:
Here we're loading the validator duties from the `head_state`, but we're producing the attestation at `present_slot`.

If the epoch of `head_state.slot` is not the same as that of `present_slot`, the duties we retrieve are not relevant to `present_slot`.

To fix this, we need the `head_state` to be skipped forward so it's aligned with `present_slot`.

I have [this method](https://github.com/sigp/lighthouse/blob/572df4f37e093c08d1c1a63d047969c5e3ee0781/beacon_node/beacon_chain/src/beacon_chain.rs#L382-L431) on the `interop` branch which does exactly that.

Reply:
Hmmm... line 212 will pick up this error.
I guess this is the kind of thing that `speculative_state` was useful for...